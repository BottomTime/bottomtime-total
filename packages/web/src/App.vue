<template>
  <div v-if="isLoading"></div>
  <div v-else>
    <SnackBar />
    <NavBar />
    <RouterView />
    <PageFooter />
    <CookieWarning />
  </div>
</template>

<script lang="ts" setup>
import { onBeforeMount, ref } from 'vue';

import CookieWarning from './components/main/CookieWarning.vue';
import { Dispatch, useStore } from '@/store';
import { inject } from '@/helpers';
import NavBar from '@/components/main/NavBar.vue';
import PageFooter from '@/components/main/PageFooter.vue';
import SnackBar from '@/components/main/SnackBar.vue';
import { ApiClientKey } from '@/injection-keys';

const client = inject(ApiClientKey);
const store = useStore();
const isLoading = ref(true);

onBeforeMount(async () => {
  try {
    const currentUser = await client.users.getCurrentUser();
    await store.dispatch(Dispatch.SignInUser, currentUser);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
});
</script>

<style>
@import './assets/bulmaswatch.min.css';
</style>
