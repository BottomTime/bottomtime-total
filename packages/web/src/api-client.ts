import { ApiClient } from '@bottomtime/api';

import { InjectionKey, inject } from 'vue';

export const ApiClientKey: InjectionKey<ApiClient> = Symbol('ApiClient');
export function useClient(): ApiClient {
  const client = inject(ApiClientKey);

  if (!client) {
    throw new Error(
      'Api Client has not been initialized. Did you forget to call `app.provide(key, client)`?',
    );
  }

  return client;
}
