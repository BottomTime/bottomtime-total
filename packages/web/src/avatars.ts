import { AvatarSize, ListAvatarURLsResponseDTO } from '@bottomtime/api';

export type AvatarURLs = ListAvatarURLsResponseDTO['sizes'];

export function getAvatarURL(
  baseUrl: string | undefined,
  size: AvatarSize,
): string | undefined {
  if (!baseUrl) return undefined;

  if (/^\/api\/users\/[a-z0-9]+([_.-][a-z0-9]+)*\/avatar/i.test(baseUrl)) {
    return baseUrl.endsWith('/') ? `${baseUrl}${size}` : `${baseUrl}/${size}`;
  }

  return baseUrl;
}
