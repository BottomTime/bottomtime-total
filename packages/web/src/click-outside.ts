import { Directive, DirectiveBinding } from 'vue';

type HTMLElementWithClickOutside = HTMLElement & {
  clickOutsideEvent: (e: MouseEvent) => void;
};

export const clickOutside: Directive = {
  beforeMount: (el: HTMLElementWithClickOutside, binding: DirectiveBinding) => {
    el.clickOutsideEvent = (event) => {
      // here I check that click was outside the el and his children
      if (!(el == event.target || el.contains(event.target as Node))) {
        // and if it did, call method provided in attribute value
        binding.value();
      }
    };
    document.addEventListener('click', el.clickOutsideEvent);
  },
  unmounted: (el: HTMLElementWithClickOutside) => {
    document.removeEventListener('click', el.clickOutsideEvent);
  },
};
