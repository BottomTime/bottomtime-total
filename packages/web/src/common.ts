import { AccountTier, LogBookSharing, ProfileDTO } from '@bottomtime/api';

import { ComputedRef } from 'vue';

export type Breadcrumb = {
  label: string | ComputedRef<string>;
  to?: string | ComputedRef<string>;
  active?: boolean;
};

export type Coordinates = {
  left: number;
  top: number;
  width: number;
  height: number;
};

export const DateTimeFormat = 'yyyy-MMM-dd hh:mm:ss a';

export const DefaultProfile: ProfileDTO = {
  accountTier: AccountTier.Basic,
  logBookSharing: LogBookSharing.Private,
  memberSince: 0,
  username: '',
  userId: '',
};

export type Selectable<T> = T & { selected?: boolean };

export type SelectOption = {
  disabled?: boolean;
  label?: string;
  value: string;
};

export type TabInfo = {
  key: string;
  label: string;
  disabled?: boolean;
  visible?: boolean;
};

export enum ToastType {
  Error = 'error',
  Info = 'info',
  Success = 'success',
  Warning = 'warning',
}

export type Toast = {
  id: string;
  message: string;
  type: ToastType;
};

export const FortyEightHoursInSeconds = 48 * 60 * 60 * 1000;
