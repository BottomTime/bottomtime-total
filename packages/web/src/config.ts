/* eslint-disable no-process-env */

export class Config {
  /** Email address for contacting support/admin. */
  static get adminEmail(): string {
    return process.env.BTWEB_VITE_ADMIN_EMAIL || 'admin@bottomti.me';
  }

  /** The application title as it should appear in the browser tab, etc. */
  static get appTitle(): string {
    return process.env.BTWEB_VITE_APP_TITLE || 'Bottom Time';
  }

  /** Base URL at which the app is listening for requests. */
  static get baseUrl(): string {
    return process.env.BTWEB_VITE_BASE_URL || 'http://localhost:4850/';
  }

  /** Whether or not to invoke Google's Places APIs. (This needs to be set to true to enable autocomplete in location boxes.) */
  static get enablePlacesApi(): boolean {
    return Boolean(process.env.BTWEB_VITE_ENABLE_PLACES_API);
  }

  /** The value of the `NODE_ENV` environment variable. */
  static get env(): string {
    return process.env.NODE_ENV || 'development';
  }

  /** Google API key for accessing the Google Maps API. */
  static get googleApiKey(): string {
    return process.env.BTWEB_VITE_GOOGLE_API_KEY || '';
  }

  /**
   * Vite mode. (`development` or `production`). This differs from `NODE_ENV`.
   * See documentation here https://vitejs.dev/guide/env-and-mode.html#node-env-and-modes
   */
  static get mode(): string {
    return process.env.MODE || 'development';
  }

  static get isProduction(): boolean {
    return Config.env === 'production';
  }

  /**
   * Detects server-side rendering.
   * Will be true if we are currently executing on the server-side or false if we are running in the browser.
   */
  static get isSSR(): boolean {
    return typeof window === 'undefined';
  }
}
