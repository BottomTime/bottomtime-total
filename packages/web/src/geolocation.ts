import { GpsCoordinates } from '@bottomtime/api';

import { InjectionKey, inject } from 'vue';

import { Logger } from './logger';

export interface IGeolocation {
  getCurrentLocation(): Promise<GpsCoordinates | undefined>;
}

export const GeolocationKey: InjectionKey<IGeolocation> = Symbol('Geolocation');

export function useGeolocation(): IGeolocation {
  const geolocation = inject(GeolocationKey);
  if (!geolocation) {
    throw new Error(
      'Geolocation not provided. Did you remember to call `app.provide()`?',
    );
  }

  return geolocation;
}

export class Geolocation implements IGeolocation {
  async getCurrentLocation(): Promise<GpsCoordinates | undefined> {
    try {
      const position = await new Promise<GeolocationPosition>(
        (resolve, reject) => {
          navigator.geolocation.getCurrentPosition(resolve, reject);
        },
      );

      return {
        lat: position.coords.latitude,
        lon: position.coords.longitude,
      };
    } catch (error) {
      Logger.warn('Failed to lookup current location', error);
      return undefined;
    }
  }
}
