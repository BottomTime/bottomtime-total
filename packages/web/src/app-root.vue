<template>
  <SnackBar />
  <div
    id="modal"
    class="font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950"
  ></div>
  <div id="drawer"></div>

  <div
    class="md:container mx-auto font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950 rounded-b-xl shadow-md shadow-white dark:shadow-grey-700 opacity-95"
  >
    <!-- Placeholder element. Modal dialogs will be "teleported" here to get the z-indexing right. -->

    <section class="mt-16 p-4">
      <RouterView></RouterView>
    </section>
    <PageFooter />
  </div>
  <NavBar />
</template>

<script setup lang="ts">
import { onBeforeMount } from 'vue';

import { ToastType } from './common';
import NavBar from './components/core/nav-bar.vue';
import PageFooter from './components/core/page-footer.vue';
import SnackBar from './components/core/snack-bar.vue';
import { Config } from './config';
import { useErrors, useToasts } from './store';

const toasts = useToasts();
const errors = useErrors();

onBeforeMount(() => {
  if (!Config.isSSR) {
    if (errors.renderError) {
      if (!Config.isProduction) {
        /* eslint-disable-next-line no-console */
        console.error(errors.renderError);
      }
      toasts.toast({
        id: 'server-prerender-error',
        message:
          'A server error has occurred and the page may not have loaded correctly. We apologize for the inconvenience. Please try again later as we investigate the issue.',
        type: ToastType.Error,
      });
    }
  }
});
</script>
