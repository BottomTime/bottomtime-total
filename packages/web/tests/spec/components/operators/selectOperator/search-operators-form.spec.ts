import {
  ApiClient,
  ApiList,
  OperatorDTO,
  SearchOperatorsResponseSchema,
  VerificationStatus,
} from '@bottomtime/api';

import {
  ComponentMountingOptions,
  flushPromises,
  mount,
} from '@vue/test-utils';

import { Pinia, createPinia } from 'pinia';

import { ApiClientKey } from '../../../../../src/api-client';
import SearchOperatorsForm from '../../../../../src/components/operators/selectOperator/search-operators-form.vue';
import SelectOperatorListItem from '../../../../../src/components/operators/selectOperator/select-operator-list-item.vue';
import { GeolocationKey } from '../../../../../src/geolocation';
import TestData from '../../../../fixtures/dive-operators.json';
import { MockGeolocation } from '../../../../mock-geolocation';

const ClearLocationButton = '#btn-clear-operator-location';
const CountsText = '[data-testid="search-operator-counts"]';
const LoadMoreButton = '#searchOperatorsLoadMore';
const NoResultsMessage = '[data-testid="search-operators-no-results"]';
const QueryInput = '#operatorSearchQuery';
const SearchButton = '#searchOperators';
const VerifiedCheckbox = '#operatorSearchVerified';

describe('SearchOperatorsForm component', () => {
  let client: ApiClient;
  let testData: ApiList<OperatorDTO>;
  let geolocation: MockGeolocation;

  let pinia: Pinia;
  let opts: ComponentMountingOptions<typeof SearchOperatorsForm>;

  beforeAll(() => {
    testData = SearchOperatorsResponseSchema.parse(TestData);
    client = new ApiClient();
    geolocation = new MockGeolocation();
  });

  beforeEach(() => {
    pinia = createPinia();
    geolocation.currentLocation = {
      lat: 43.646,
      lon: -79.384,
    };
    opts = {
      global: {
        provide: {
          [ApiClientKey as symbol]: client,
          [GeolocationKey as symbol]: geolocation,
        },
        plugins: [pinia],
      },
    };
  });

  it("will perform a basic search based on user's current location", async () => {
    const spy = jest
      .spyOn(client.operators, 'searchOperators')
      .mockResolvedValueOnce(testData);
    const wrapper = mount(SearchOperatorsForm, opts);
    await flushPromises();

    await wrapper.get(SearchButton).trigger('click');
    await flushPromises();

    expect(spy).toHaveBeenCalledWith({
      location: geolocation.currentLocation,
      radius: 50,
    });

    expect(wrapper.get(CountsText).text()).toMatchSnapshot();
    expect(wrapper.get(LoadMoreButton).isVisible()).toBe(true);

    const items = wrapper.findAllComponents(SelectOperatorListItem);
    expect(items).toHaveLength(testData.data.length);
    items.forEach((item, index) => {
      expect(item.props('operator')).toEqual(testData.data[index]);
    });
  });

  it("will perform a basic search if user's location cannot be determined", async () => {
    geolocation.currentLocation = undefined;
    const spy = jest
      .spyOn(client.operators, 'searchOperators')
      .mockResolvedValueOnce(testData);
    const wrapper = mount(SearchOperatorsForm, opts);
    await flushPromises();

    await wrapper.get(SearchButton).trigger('click');
    await flushPromises();

    expect(spy).toHaveBeenCalledWith({});

    expect(wrapper.get(CountsText).text()).toMatchSnapshot();
    expect(wrapper.get(LoadMoreButton).isVisible()).toBe(true);

    const items = wrapper.findAllComponents(SelectOperatorListItem);
    expect(items).toHaveLength(testData.data.length);
    items.forEach((item, index) => {
      expect(item.props('operator')).toEqual(testData.data[index]);
    });
  });

  it('will perform a more robust search', async () => {
    const query = 'best dive shop';
    const spy = jest
      .spyOn(client.operators, 'searchOperators')
      .mockResolvedValueOnce(testData);
    const wrapper = mount(SearchOperatorsForm, opts);
    await flushPromises();

    await wrapper.get(QueryInput).setValue(query);
    await wrapper.get(VerifiedCheckbox).setValue(true);
    await wrapper.get(SearchButton).trigger('click');
    await flushPromises();

    expect(spy).toHaveBeenCalledWith({
      query,
      verification: VerificationStatus.Verified,
      location: geolocation.currentLocation,
      radius: 50,
    });

    expect(wrapper.get(CountsText).text()).toMatchSnapshot();
    expect(wrapper.get(LoadMoreButton).isVisible()).toBe(true);

    const items = wrapper.findAllComponents(SelectOperatorListItem);
    expect(items).toHaveLength(testData.data.length);
    items.forEach((item, index) => {
      expect(item.props('operator')).toEqual(testData.data[index]);
    });
  });

  it('will load more results on request', async () => {
    const spy = jest
      .spyOn(client.operators, 'searchOperators')
      .mockResolvedValueOnce({
        data: testData.data.slice(0, 10),
        totalCount: testData.totalCount,
      })
      .mockResolvedValueOnce({
        data: testData.data.slice(10, 20),
        totalCount: testData.totalCount,
      });
    const wrapper = mount(SearchOperatorsForm, opts);
    await flushPromises();

    await wrapper.get(SearchButton).trigger('click');
    await flushPromises();

    await wrapper.get(LoadMoreButton).trigger('click');
    await flushPromises();

    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy).toHaveBeenCalledWith({
      location: geolocation.currentLocation,
      radius: 50,
    });
    expect(spy).toHaveBeenCalledWith({
      location: geolocation.currentLocation,
      radius: 50,
      skip: 10,
    });

    expect(wrapper.get(CountsText).text()).toMatchSnapshot();

    const items = wrapper.findAllComponents(SelectOperatorListItem);
    expect(items).toHaveLength(20);
    items.forEach((item, index) => {
      expect(item.props('operator')).toEqual(testData.data[index]);
    });
  });

  it('will display a message if there are no results to show', async () => {
    const spy = jest
      .spyOn(client.operators, 'searchOperators')
      .mockResolvedValueOnce({ data: [], totalCount: 0 });
    const wrapper = mount(SearchOperatorsForm, opts);
    await flushPromises();

    await wrapper.get(SearchButton).trigger('click');
    await flushPromises();

    expect(spy).toHaveBeenCalledWith({
      location: geolocation.currentLocation,
      radius: 50,
    });

    expect(wrapper.get(CountsText).text()).toMatchSnapshot();
    expect(wrapper.find(LoadMoreButton).exists()).toBe(false);
    expect(wrapper.get(NoResultsMessage).isVisible()).toBe(true);
  });

  it('will allow the user to clear the location', async () => {
    const spy = jest
      .spyOn(client.operators, 'searchOperators')
      .mockResolvedValueOnce(testData);
    const wrapper = mount(SearchOperatorsForm, opts);
    await flushPromises();

    await wrapper.get(ClearLocationButton).trigger('click');
    await wrapper.get(SearchButton).trigger('click');

    expect(spy).toHaveBeenCalledWith({});
  });
});
