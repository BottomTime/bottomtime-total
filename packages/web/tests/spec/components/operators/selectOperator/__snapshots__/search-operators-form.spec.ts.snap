// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchOperatorsForm component will display a message if there are no results to show 1`] = `"Showing 0 of 0 results."`;

exports[`SearchOperatorsForm component will load more results on request 1`] = `"Showing 20 of 651 results."`;

exports[`SearchOperatorsForm component will perform a basic search based on user's current location 1`] = `"Showing 49 of 651 results."`;

exports[`SearchOperatorsForm component will perform a basic search if user's location cannot be determined 1`] = `"Showing 49 of 651 results."`;

exports[`SearchOperatorsForm component will perform a more robust search 1`] = `"Showing 49 of 651 results."`;
